cdap-watchdog/src/main/java/co/cask/cdap/logging/gateway/handlers/AbstractChunkedLogProducer.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.logging.gateway.handlers;
+
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.logging.read.LogEvent;
+import co.cask.http.BodyProducer;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Multimap;
+import org.jboss.netty.buffer.ChannelBuffer;
+import org.jboss.netty.buffer.ChannelBuffers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.Charset;
+import java.nio.charset.CharsetEncoder;
+import java.nio.charset.CoderResult;
+import javax.annotation.Nullable;
+
+/**
+ * LogReader BodyProducer class that delegates to subclasses for how to encode log events.
+ */
+public abstract class AbstractChunkedLogProducer extends BodyProducer {
```

Comment:
This class looks unnecessarily complicated. I think the contract of this class is to provide the skeleton to loop over an iterator of `LogEvent`. It should be as simple as follow:

```java
public abstract class AbstractLogBodyProducer extends BodyProducer {

  private final CloseableIterator<LogEvent> logEventIterator;
  private boolean hasStarted = false;
  private boolean hasFinished = false;

  protected AbstractLogBodyProducer(CloseableIterator<LogEvent> logEventIterator) {
    this.logEventIterator = logEventIterator;
  }

  protected abstract ChannelBuffer onWriteStart() throws IOException;
  protected abstract ChannelBuffer writeLogEvent(LogEvent logEvent) throws IOException;
  protected abstract ChannelBuffer onWriteFinish() throws IOException;

  @Override
  public final ChannelBuffer nextChunk() throws Exception {
    if (!hasStarted) {
      hasStarted = true;
      return onWriteStart();
    }
    if (logEventIterator.hasNext()) {
      return writeLogEvent(logEventIterator.next());
    }
    if (!hasFinished) {
      hasFinished = true;
      return onWriteFinish();
    }

    return ChannelBuffers.EMPTY_BUFFER;
  }

  @Override
  public void finished() throws Exception {
    logEventIterator.close();
  }

  @Override
  public final void handleError(@Nullable Throwable throwable) {
    // ...
  }
}
```

Then the sub-class is responsible for turning `LogEvent` into `ChannelBuffer`. (I just give a simple one here, but you can make it more complicate for the field filtering purpose).

```java
public class SimpleJsonLogBodyProducer extends AbstractLogBodyProducer {

  private static final Gson GSON = new Gson();

  private final ChannelBuffer channelBuffer;
  private final JsonWriter jsonWriter;

  protected SimpleJsonLogBodyProducer(CloseableIterator<LogEvent> logEventIterator) {
    super(logEventIterator);
    this.channelBuffer = ChannelBuffers.dynamicBuffer(8192);
    this.jsonWriter = new JsonWriter(new OutputStreamWriter(new ChannelBufferOutputStream(channelBuffer),
                                                            StandardCharsets.UTF_8));
  }

  @Override
  protected ChannelBuffer onWriteStart() throws IOException {
    channelBuffer.clear();
    jsonWriter.beginArray();
    jsonWriter.flush();
    return channelBuffer;
  }

  @Override
  protected ChannelBuffer writeLogEvent(LogEvent logEvent) throws IOException {
    channelBuffer.clear();
    GSON.toJson(logEvent, LogEvent.class, jsonWriter);
    jsonWriter.flush();
    return channelBuffer;
  }

  @Override
  protected ChannelBuffer onWriteFinish() throws IOException {
    channelBuffer.clear();
    jsonWriter.endArray();
    jsonWriter.flush();
    return channelBuffer;
  }
}
```